man/ibping.8
.TH IBPING 8 "August 11, 2006" "OpenIB" "OpenIB Diagnostics"

.SH NAME
ibping \- ping an InfiniBand address

.SH SYNOPSIS
.B ibping
[\-d(ebug)] [\-e(rr_show)] [\-v(erbose)] [\-G(uid)] [\-C ca_name]
[\-P ca_port] [\-s smlid] [\-t(imeout) timeout_ms] [\-V(ersion)]
[\-c ping_count] [\-f(lood)] [\-o oui] [\-S(erver)] [\-h(elp)]
<dest lid | guid>

.SH DESCRIPTION
.PP
ibping uses vendor mads to validate connectivity between IB nodes.
On exit, (IP) ping like output is show. ibping is run as client/server.
Default is to run as client. Note also that a default ping server is
implemented within the kernel.

.SH OPTIONS

.PP
.TP
\fB\-c\fR
stop after count packets
.TP
\fB\-f\fR, \fB\-\-flood\fR
flood destination: send packets back to back without delay
.TP
\fB\-o\fR, \fB\-\-oui\fR
use specified OUI number to multiplex vendor mads
.TP
\fB\-S\fR, \fB\-\-Server\fR
start in server mode (do not return)

.SH COMMON OPTIONS

Most OpenIB diagnostics take the following common flags. The exact list of
supported flags per utility can be found in the usage message and can be shown
using the util_name -h syntax.

# Debugging flags
.PP
\-d      raise the IB debugging level.
        May be used several times (-ddd or -d -d -d).
.PP
\-e      show send and receive errors (timeouts and others)
.PP
\-h      show the usage message
.PP
\-v      increase the application verbosity level.
        May be used several times (-vv or -v -v -v)
.PP
\-V      show the version info.

# Addressing flags
.PP
\-G      use GUID address argument. In most cases, it is the Port GUID.
        Example:
        "0x08f1040023"
.PP
\-s <smlid>      use 'smlid' as the target lid for SM/SA queries.

# Other common flags:
.PP
\-C <ca_name>    use the specified ca_name.
.PP
\-P <ca_port>    use the specified ca_port.
.PP
\-t <timeout_ms> override the default timeout for the solicited mads.

Multiple CA/Multiple Port Support

When no IB device or port is specified, the port to use is selected
by the following criteria:
.PP
1. the first port that is ACTIVE.
.PP
2. if not found, the first port that is UP (physical link up).

If a port and/or CA name is specified, the user request is
attempted to be fulfilled, and will fail if it is not possible.

.SH AUTHOR
.TP
Hal Rosenstock
.RI < halr@voltaire.com >


ChangeLog
commit 596aa88065f89906884a821a1bb4eb631abd6af0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:57:58 2009 -0500

    Spin release version 1.2.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0ec532c2b6b92be6adf58849c9e3d215a50d306d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:52:51 2009 -0500

    Support new device ids: 0x35, 0x36.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f685c8fe7e77e64614d825e563dd9f02a0b1ae16
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:46 2008 -0500

    Spin release 1.2.2
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ebd19d077d331270805aefa9ca00d3661561ba87
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:27 2008 -0500

    Relax the required firmware level.
    
    Turns out that libcxgb3 can really run on either 6.0 or 5.0
    firmware.  So I'm relaxing the requirement...
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 6f7485feb244d8571fcab2292ef92c97bea48df0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue May 13 09:09:35 2008 -0500

    Spin release 1.2.1.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 309916b23ba384a0bd297e7e51b6b89bff357a78
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:45:01 2008 -0500

    Use local stag == 1 to silently drop the p2p read response.
    
    The current code fails if there are read completions -and-
    the p2p read response in the cq when its flushed.
    
    This fix uses the fact that the RTR read is to target stag 1,
    which is not a valid stag.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 67d519c4a9e28e54a9e436404156c9ae7eb60314
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:25:41 2008 -0500

    Handle wrapping when flushing the software sq.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a3dcea751a94ad5e0a9cad6c55a839a9848f327c
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Spin 1.2.0 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit cc2ef741842883233a77e18d65a828c711e581a3
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit 92322b93f48fa98b4fadb736ae49dafda237272a
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Apr 24 14:23:50 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit a8fb6221af4d914f39c1e71f9331f6d428d698c2
Author: Steve Wise (Open Grid Computing) <swise@hosting.openfabrics.org>
Date:   Tue Apr 1 15:30:11 2008 -0700

    Spin 1.1.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit db41ebc10f4cc4c2fe7826589636ab7d4b804218
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 26 08:27:12 2008 -0600

    Provide an empty stub for iwch_query_qp.
    
    This prevents crashes when calling ibv_query_qp due to an expected
    library provided function.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ec4f7f1c6b8beceb9001519bc4e2ee89fbb2e6d0
Author: Dotan Barak <dotanb@dev.mellanox.co.il>
Date:   Tue Feb 26 08:23:36 2008 -0600

    Remove duplicated line.
    
    Signed-off-by: Dotan Barak <dotanb@dev.mellanox.co.il>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 8218482d03898f859e1fa272649fafc64f91f05f
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Feb 12 15:16:03 2008 -0600

    Spin release 1.1.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 1bc6930d8884646ef73db71a95cc32c96c20ba32
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 12 14:46:10 2008 -0600

    libcxgb3: zeroing of wc_flags
    
    The wc_flags field in struct ibv_wc is left uninitialized in
    iwch_poll_cq_one.  User space applications may check this field and
    deterministically perform actions based on the garbage in the field.
    Zeroing this out will prevent this unintended behavior.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 946e1572d38d786038d3bca5a2ec184dcf7636c4
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jan 25 16:47:01 2008 -0600

    Spin release 1.1.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 94fb771a44212a679d7f83ee69df24e841e0ce08
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jan 25 13:04:31 2008 -0600

    Fix path for obtaining the firmware version.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 4bbb15965b53cad6273094e95b06895010ca025c
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Wed Jan 23 09:40:28 2008 -0600

    Non-supported functions should return NULL when returning pointers.
    
    Some/Most user space programs will not check for a (void *) to -ENOSYS,
    which can look like a real address until referenced.
    
    This patch converts the uses of (void *) -ENOSYS to NULL.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 10893a7a45e6913483d90023bb90bf9cb2420384
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jan 3 17:19:27 2008 -0600

    Spin release 1.1.2.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a35c74c32d038bd7508ad3cb6c5291c1865d3ba4
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jan 3 17:16:27 2008 -0600

    Fix compiler warnings on 32b.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ef3b9d4f99eb6e18c29b70599edf99d0715f03fb

src/ibping.c
/*
 * Copyright (c) 2004-2008 Voltaire Inc.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <getopt.h>

#include <infiniband/umad.h>
#include <infiniband/mad.h>
#include <complib/cl_timer.h>

#include "ibdiag_common.h"

struct ibmad_port *srcport;

static char host_and_domain[IB_VENDOR_RANGE2_DATA_SIZE];
static char last_host[IB_VENDOR_RANGE2_DATA_SIZE];

static void
get_host_and_domain(char *data, int sz)
{
	char *s = data;
	int n;

	if (gethostname(s, sz) < 0)
		snprintf(s, sz, "?hostname?");

	s[sz-1] = 0;
	if ((n = strlen(s)) >= sz)
		return;
	s[n] = '.';
	s += n + 1;
	sz -= n + 1;

	if (getdomainname(s, sz) < 0)
		snprintf(s, sz, "?domainname?");
	if (strlen(s) == 0)
		s[-1] = 0;	/* no domain */
}

static char *
ibping_serv(void)
{
	void *umad;
	void *mad;
	char *data;

	DEBUG("starting to serve...");

	while ((umad = mad_receive_via(0, -1, srcport))) {

		mad = umad_get_mad(umad);
		data = (char *)mad + IB_VENDOR_RANGE2_DATA_OFFS;

		memcpy(data, host_and_domain, IB_VENDOR_RANGE2_DATA_SIZE);

		DEBUG("Pong: %s", data);

		if (mad_respond_via(umad, 0, 0, srcport) < 0)
			DEBUG("respond failed");

		mad_free(umad);
	}

	DEBUG("server out");
	return 0;
}

static uint64_t
ibping(ib_portid_t *portid, int quiet)
{
	char data[IB_VENDOR_RANGE2_DATA_SIZE] = {0};
	ib_vendor_call_t call;
	uint64_t start, rtt;

	DEBUG("Ping..");

	start = cl_get_time_stamp();

	call.method = IB_MAD_METHOD_GET;
	call.mgmt_class = IB_VENDOR_OPENIB_PING_CLASS;
	call.attrid = 0;
	call.mod = 0;
	call.oui = IB_OPENIB_OUI;
	call.timeout = 0;
	memset(&call.rmpp, 0, sizeof call.rmpp);

	if (!ib_vendor_call_via(data, portid, &call, srcport))
		return ~0ull;

	rtt = cl_get_time_stamp() - start;

	if (!last_host[0])
		memcpy(last_host, data, sizeof last_host);

	if (!quiet)
		printf("Pong from %s (%s): time %" PRIu64 ".%03" PRIu64 " ms\n",
			data, portid2str(portid), rtt/1000, rtt%1000);

	return rtt;
}

static uint64_t minrtt = ~0ull, maxrtt, total_rtt;
static uint64_t start, total_time, replied, lost, ntrans;
static ib_portid_t portid = {0};

void report(int sig)
{
	total_time = cl_get_time_stamp() - start;

	DEBUG("out due signal %d", sig);

	printf("\n--- %s (%s) ibping statistics ---\n", last_host, portid2str(&portid));
	printf("%" PRIu64 " packets transmitted, %" PRIu64 " received, %" PRIu64 "%% packet loss, time %" PRIu64 " ms\n",
		ntrans, replied,
		(lost != 0) ?  lost * 100 / ntrans : 0, total_time / 1000);
	printf("rtt min/avg/max = %" PRIu64 ".%03" PRIu64 "/%" PRIu64 ".%03" PRIu64 "/%" PRIu64 ".%03" PRIu64 " ms\n",
		minrtt == ~0ull ? 0 : minrtt/1000,
		minrtt == ~0ull ? 0 : minrtt%1000,
		replied ? total_rtt/replied/1000 : 0,
		replied ? (total_rtt/replied)%1000 : 0,
		maxrtt/1000, maxrtt%1000);

	exit(0);
}

static int server = 0, flood = 0, oui = IB_OPENIB_OUI;
static unsigned count = ~0;

static int process_opt(void *context, int ch, char *optarg)
{
	switch (ch) {
	case 'c':
		count = strtoul(optarg, 0, 0);
		break;
	case 'f':
		flood++;
		break;
	case 'o':
		oui = strtoul(optarg, 0, 0);
		break;
	case 'S':
		server++;
		break;
	default:
		return -1;
	}
	return 0;
}

int main(int argc, char **argv)
{
	int mgmt_classes[3] = {IB_SMI_CLASS, IB_SMI_DIRECT_CLASS, IB_SA_CLASS};
	int ping_class = IB_VENDOR_OPENIB_PING_CLASS;
	uint64_t rtt;
	char *err;

	const struct ibdiag_opt opts[] = {
		{ "count", 'c', 1, "<num>", "stop after count packets" },
		{ "flood", 'f', 0, NULL, "flood destination" },
		{ "oui", 'o', 1, NULL, "use specified OUI number" },
		{ "Server", 'S', 0, NULL, "start in server mode" },
		{ 0 }
	};
	char usage_args[] = "<dest lid|guid>";

	ibdiag_process_opts(argc, argv, NULL, "D", opts, process_opt,
			    usage_args, NULL);

	argc -= optind;
	argv += optind;

	if (!argc && !server)
		ibdiag_show_usage();

	srcport = mad_rpc_open_port(ibd_ca, ibd_ca_port, mgmt_classes, 3);
	if (!srcport)
		IBERROR("Failed to open '%s' port '%d'", ibd_ca, ibd_ca_port);

	if (server) {
		if (mad_register_server_via(ping_class, 0, 0, oui, srcport) < 0)
			IBERROR("can't serve class %d on this port", ping_class);

		get_host_and_domain(host_and_domain, sizeof host_and_domain);

		if ((err = ibping_serv()))
			IBERROR("ibping to %s: %s", portid2str(&portid), err);
		exit(0);
	}

	if (mad_register_client_via(ping_class, 0, srcport) < 0)
		IBERROR("can't register ping class %d on this port", ping_class);

	if (ib_resolve_portid_str_via(&portid, argv[0], ibd_dest_type,
					ibd_sm_id, srcport) < 0)
		IBERROR("can't resolve destination port %s", argv[0]);

	signal(SIGINT, report);
	signal(SIGTERM, report);

	start = cl_get_time_stamp();

	while (count-- > 0) {
		ntrans++;
		if ((rtt = ibping(&portid, flood)) == ~0ull) {
			DEBUG("ibping to %s failed", portid2str(&portid));
			lost++;
		} else {
			if (rtt < minrtt)
				minrtt = rtt;
			if (rtt > maxrtt)
				maxrtt = rtt;
			total_rtt += rtt;
			replied++;
		}

		if (!flood)
			sleep(1);
	}

	report(0);

	mad_rpc_close_port(srcport);

	exit(-1);
}
